### PR TITLE
Loading custom names for lamps

### PR DESCRIPTION
**Pull request by kbacia7**
Hi, I added support for custom names of lamps.
In `devices.yaml` we can define names for lamps by short address, like:
```yaml
12: "lampInKitchen"
```

And later in HA/MQTT will can use custom name "lampInKitchen"
I added loading custom names in function `initialize_lamps`, because is possible to reload lamps and names without restarting script

### CODE REVIEW

File: devicesnamesconfig.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+            open(self._path, "w")
+            self._devices_names = {}
+
+    def load_devices_names_file(self):
+        """Load configuration from yaml file."""
+        with open(self._path, "r") as infile:
+            logger.debug("Loading devices names from <%s>", self._path)
+            try:
+                self._devices_names = yaml.safe_load(infile)
+            except vol.MultipleInvalid as error:
```
Comment (dgomes) [inline devicesnamesconfig.py:25]: are you sure about this exception ? is this something `yaml` library throws ?

File: devicesnamesconfig.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+                logger.error("In configuration file %s: %s", self._path, error)
+                quit(1)
```
Comment (dgomes) [inline devicesnamesconfig.py:27]: raise a new error, don't abort from a support class

File: dali_mqtt_daemon.py [hunk 3]
```diff
@@ -91,20 +95,30 @@ def initialize_lamps(data_object, client):
...
             physical_minimum = driver_object.send(
                 gear.QueryPhysicalMinimum(short_address)
             )
             min_level = driver_object.send(gear.QueryMinLevel(short_address))
             max_level = driver_object.send(gear.QueryMaxLevel(short_address))
+            device_name = None
+            try:
+                device_name = devices_names[short_address_value]
+            except KeyError:
+                device_name = short_address_value
```
Comment (dgomes) [inline dali_mqtt_daemon.py:109]: This should be handled by the DevicesNamesConfig or even by Lamp

**Review by dgomes — CHANGES_REQUESTED** [review]
I like the idea a lot!

**Comment by kbacia7** [comment]
All fixed, if you don't have more suggestions about code please wait a moment with merge this PR.
I was talking with Adam and we came to the conclusion that `devices.yaml` should have [syntax like devices names have in  zigbee2mqt](https://www.zigbee2mqtt.io/information/configuration.html#specifying-devices-and-groups-in-a-separate-file)

File: devicesnamesconfig.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+class DevicesNamesConfig:
+    def __init__(self):
+        self._path = "devices.yaml"
+        self._devices_names = {}
+        # Load from file
+        try:
+            self.load_devices_names_file()
+        except FileNotFoundError:
+            logger.info("No device names config, creating new one")
+            open(self._path, "w")
```
Comment (dgomes) [inline devicesnamesconfig.py:20]: use "with" to create a context in order to make sure the file is closed

**Comment by kbacia7** [comment]
Done, now syntax `devices.yaml` is:
```yaml
12: 
   "friendly_name": "lampInKitchen"
```

And filename is passed via argument or saved in `config.yaml`

**Comment by kbacia7** [comment]
A few more changes:
* Fix a bug with crash on sending switch OFF
* Reconize friendly name and device name like zigbee2mqt - friendly name is from `devices.yaml` and is showed in HA, device name is a snake case version of friendly name and is used in MQTT, if device didn't have friendly name will be used short address
* Removed `dali2mqtt_` prefix from lamps config and HA discovery path
Tell me what do you think about this changes

File: config.py [hunk 5]
```diff
@@ -5,18 +5,15 @@
 from watchdog.events import FileSystemEventHandler
 from watchdog.observers.polling import PollingObserver as Observer
 
-from consts import RESET_COLOR
-
-log_format = "%(asctime)s %(levelname)s: %(message)s{}".format(RESET_COLOR)
-logging.basicConfig(format=log_format)
-logger = logging.getLogger(__name__)
+from consts import logger
```
Comment (dgomes) [inline config.py:8]: Can't do this... `__name__` will always be "consts" and we would like to know which module produced the log

File: dali_mqtt_daemon.py [hunk 6]
```diff
@@ -91,21 +96,29 @@ def initialize_lamps(data_object, client):
     for lamp in lamps:
         try:
             short_address = address.Short(lamp)
+            short_address_value = short_address.address
```
Comment (dgomes) [inline dali_mqtt_daemon.py:99]: do we really need this extra variable ?

File: devicesnamesconfig.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        try:
+            self.load_devices_names_file()
+        except FileNotFoundError:
+            logger.info("No device names config, creating new one")
+            with open(self._path, "w"):
+                pass
+
+    def load_devices_names_file(self):
+        """Load configuration from yaml file."""
+        with open(self._path, "r") as infile:
```
Comment (dgomes) [inline devicesnamesconfig.py:25]: move with inside try

File: devicesnamesconfig.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+                logger.error("In devices file %s: %s", self._path, error)
+                raise DevicesNamesConfigLoadError()
+
+    def save_devices_names_file(self, all_lamps):
+        self._devices_names = {}
+        for lamp_object in all_lamps.values():
+            self._devices_names[lamp_object.short_address.address] = {
+                "friendly_name": str(lamp_object.short_address.address)
+            }
+        with open(self._path, "w") as outfile:
```
Comment (dgomes) [inline devicesnamesconfig.py:39]: add try ... except

File: lamp.py [hunk 9]
```diff
@@ -22,10 +22,19 @@
...
+        short_address,
+        min_physical_level,
+        min_level,
+        level,
+        max_level,
     ):
         self.driver = driver
         self.short_address = short_address
+        self.friendly_name = friendly_name
+        self.device_name = friendly_name.replace(" ", "_").lower()
```
Comment (dgomes) [inline lamp.py:37]: Use https://github.com/un33k/python-slugify

**Review by dgomes — CHANGES_REQUESTED** [review]
This has become a very large PR... lets try to keep these small (easier to review)

File: config.py (hunk 5 above)
Comment (kbacia7) [reply config.py:8]: What I should do? We probably don't want create in every module new logger object and load config to set correct level (#19)

Comment (dgomes) [reply config.py:8]: Each model should have it's own logger object, that's the common way

**Comment by dgomes** [comment]
We need to document this in the README.md

**Comment by kbacia7** [comment]
> We need to document this in the README.md

Done

File: README.md [hunk 10]
```diff
@@ -39,7 +40,22 @@ Then just edit the file accordingly. You can also create the file with the right
...
   --ha-discover-prefix HA_DISCOVER_PREFIX
                         HA discover mqtt prefix
+  --log-level {critical,error,warning,info,debug}  
+                        Log level  
+  --log-color Coloring output
+```
+
+### Devices friendly names
+Default all lamps will be displayed in Home Assistant by short address, numbers from 0 to 63
+You can give lamps special names to identify lamps by name. After first run of script, `devices.yaml` with all lamps will created.
```
Comment (dgomes) [inline README.md:50]:
```suggestion
You can give lamps special names to help you identify lamps by name. On the first execution, `devices.yaml` file will be create with all lamps available.
```

File: README.md [hunk 11]
```diff
@@ -39,7 +40,22 @@ Then just edit the file accordingly. You can also create the file with the right
...
+Example of filled `devices.yaml`:
```
Comment (dgomes) [inline README.md:51]:
```suggestion
Example `devices.yaml`:
```

File: README.md [hunk 12]
```diff
@@ -39,7 +40,22 @@ Then just edit the file accordingly. You can also create the file with the right
...
+```yaml
+0: 
+  "friendly_name": "Lamp in kitchen"
+8:
+  "friendly_name": "Lamp in bathroom"
 ```
+If you want use lamps by post messages on MQTT, remember that friendly names are converted to slug strings, so lamp with address 0 (from example) in MQTT will be named "lamp-in-kitchen"
```
Comment (dgomes) [inline README.md:58]:
```suggestion
Please note that MQTT topics support a minimum set of characters, therefore friendly names are converted to slug strings, so a lamp with address 0 (as an example) in MQTT will be named "lamp-in-kitchen"
```

**Review by dgomes — CHANGES_REQUESTED** [review]
move the logger back to each class

File: config.py (hunk 5 above)
Comment (kbacia7) [reply config.py:8]: So I should pass `log_level` to every class to set correct log level?

**Review by dgomes — APPROVED** [review]

**Review by dgomes — APPROVED** [review]

**Pull request merged**